### PR TITLE
Added browser refresh button support

### PR DESCRIPTION
**Pull request by svenjens**
## Please Squash and Merge instead of normal Merge

image

### CODE REVIEW

File: apps/client/src/context.js
Comment (afjlambert) [inline apps/client/src/context.js]: `defaultValues` is no longer used outside of this file right?
Please remove the export if we can.

File: apps/client/src/context.js
Comment (afjlambert) [inline apps/client/src/context.js:1]: Can you add some comments in this file please? For me it's not really clear what we are actually doing here.

File: apps/client/src/context.js
Comment (afjlambert) [inline apps/client/src/context.js]: Can we give a more descriptive name, maybe SessionProvider?

File: apps/client/src/context.js
Comment (afjlambert) [inline apps/client/src/context.js]: Can we move this up to the top? It's used in the `reducer` function.

File: apps/client/src/context.js
Comment (afjlambert) [inline apps/client/src/context.js]: Can this be a const?

File: apps/client/src/context.js
Comment (afjlambert) [inline apps/client/src/context.js]: Can we use an English variable name here? Also we never use camelCase for "vergunningcheck" as far as I know.

File: apps/client/src/context.js
Comment (afjlambert) [inline apps/client/src/context.js]: Good place to add a comment here?

File: apps/client/src/pages/LocationPage.js
Comment (afjlambert) [inline apps/client/src/pages/LocationPage.js]: I'm confused. In `defaultValues` there is `{ topic: null, address: null }`. There is no `data` prop.
Here we have a `setData(data)` where `data` has a `data` property.

File: apps/client/src/sttr_client/models/checker.js
Comment (afjlambert) [inline apps/client/src/sttr_client/models/checker.js]: This function returns an object with "question.id" as keys and "answer" values.
So `string: (boolean|string|number|string[])` I guess.

File: apps/client/src/pages/QuestionsPage.js
Comment (afjlambert) [inline apps/client/src/pages/QuestionsPage.js]: Can you add a comment why you need to set the data on context before you set the answer to the question. Is it because of position on the stack?

File: apps/client/src/context.js
Comment (afjlambert) [inline apps/client/src/context.js:1]: Maybe context.js is a good candidate to convert to context.ts?

**Review by afjlambert — CHANGES_REQUESTED** [review]
Looks good @svenjens.
Couple of questions though.

File: apps/client/src/sttr_client/models/checker.js
Comment (robinpiets) [inline apps/client/src/sttr_client/models/checker.js]: Shall we rename this to `getQuestionAnswers()`? @svenjens @afjlambert 

This is too general and we're probably going to need more of these getters/setters in the near future

File: apps/client/src/sttr_client/models/checker.test.js
Comment (robinpiets) [inline apps/client/src/sttr_client/models/checker.test.js:37]: I would prefer it if we change these rules to something more general and understanding, like `Answer option 1` and `Question title 1`

File: apps/client/src/pages/QuestionsPage.js
Comment (robinpiets) [inline apps/client/src/pages/QuestionsPage.js]: Can you start with the reordering of imports? And also remove the unused import?

File: apps/client/src/context.js
Comment (robinpiets) [inline apps/client/src/context.js]: typo: avaible

File: apps/client/src/context.js
Comment (robinpiets) [inline apps/client/src/context.js]: np: sessionStore > sessionStorage

File: apps/client/src/hoc/withAddress.js
Comment (robinpiets) [inline apps/client/src/hoc/withAddress.js]: Cool!!!

File: apps/client/src/pages/QuestionsPage.js
Comment (robinpiets) [inline apps/client/src/pages/QuestionsPage.js]: This is also a good place for a comment

File: apps/client/src/pages/ResultsPage.js
Comment (robinpiets) [inline apps/client/src/pages/ResultsPage.js]: Can we move the `Helmet` import up?

File: apps/client/src/pages/ResultsPage.js
Comment (robinpiets) [inline apps/client/src/pages/ResultsPage.js]: Better to rename `index` to `questionIndex`

File: apps/client/src/sttr_client/models/checker.js
Comment (robinpiets) [inline apps/client/src/sttr_client/models/checker.js]: Shall we rename this to `setQuestionAnswers()`? @svenjens @afjlambert 

This is too general and we're probably going to need more of these getters/setters in the near future

File: apps/client/src/utils/test-utils.js
Comment (robinpiets) [inline apps/client/src/utils/test-utils.js:7]: np: add empty line between external and internal imports

**Review by robinpiets — CHANGES_REQUESTED** [review]

File: apps/client/src/pages/QuestionsPage.js
Comment (svenjens) [reply apps/client/src/pages/QuestionsPage.js]: i agree

File: apps/client/src/hoc/withChecker.js
Comment (robinpiets) [inline apps/client/src/hoc/withChecker.js]: @afjlambert How to do reset the checker?

File: apps/client/src/sttr_client/models/checker.js
Comment (svenjens) [inline apps/client/src/sttr_client/models/checker.js]: 😂

File: apps/client/src/sttr_client/models/checker.test.js
Comment (svenjens) [inline apps/client/src/sttr_client/models/checker.test.js:68]: 🙌

File: apps/client/src/pages/LocationPage.js
Comment (svenjens) [inline apps/client/src/pages/LocationPage.js]: I like to see that the data is from the context. 
But i think to best practice is too destructure everything. What do you think?

Comment (robinpiets) [reply apps/client/src/pages/LocationPage.js]: Fixed. Can you review?

**Review by robinpiets — APPROVED** [review]

File: apps/client/src/context.js
Comment (afjlambert) [reply apps/client/src/context.js:1]: I though you already started working on this. Or not?
Converting this to typescript would also 'fix' the `as any` in consumers of the Contexts.

File: apps/client/src/components/QuestionAnswerTable.js
Comment (afjlambert) [inline apps/client/src/components/QuestionAnswerTable.js]: Why put question marks everywhere? Checker is guaranteed to be here right? Stack is always available on checker.

File: apps/client/src/components/QuestionAnswerTable.js
Comment (afjlambert) [inline apps/client/src/components/QuestionAnswerTable.js]: Why filter only these questions? What are you not interested in? Consider implementing an `else` case. Even if it's only a warning/notise that `type x` is not supported.

File: apps/client/src/components/QuestionAnswerTable.js
Comment (afjlambert) [inline apps/client/src/components/QuestionAnswerTable.js]: We should consider adding an atom for Icon if it's not safe to use without browser-hacks.
It's too easy to forget to implement such a 'fix'.
Not implemented in this change, I know.

File: apps/client/src/components/QuestionAnswerTable.js
Comment (afjlambert) [inline apps/client/src/components/QuestionAnswerTable.js]: isDecisiveForPermits sounds like a bool but it's actually an array :-/
Not implemented in this change, I know.

File: apps/client/src/hoc/withAddress.js
Comment (afjlambert) [reply apps/client/src/hoc/withAddress.js]: Why the question-mark? Can't we either guarantee it's an Object by declaring it above like `const { address } = useContext(SessionContext) || {};` or make sure it's added in the default session context?

File: apps/client/src/hoc/withChecker.js
Comment (afjlambert) [inline apps/client/src/hoc/withChecker.js]: But sessionContext.address is actually an Object right? Shouldn't you check for `[topic.slug]` as well?

File: apps/client/src/index.js
Comment (afjlambert) [inline apps/client/src/index.js]: Huh? What does this have to do with refresh-button?
This has never been here? Why add it now?

File: apps/client/src/pages/ConclusionPage.js
Comment (afjlambert) [inline apps/client/src/pages/ConclusionPage.js]: Again, not sure why/when address is an map with an address per topic vs just the address itself.

File: apps/client/src/pages/LocationPage.js
Comment (afjlambert) [inline apps/client/src/pages/LocationPage.js]: Again, not sure. Also question mark again.

File: apps/client/src/pages/LocationPage.js
Comment (afjlambert) [inline apps/client/src/pages/LocationPage.js]: I think we should consider if this logic should be part of the LocationPage.

File: apps/client/src/pages/LocationPage.js
Comment (afjlambert) [inline apps/client/src/pages/LocationPage.js]: Can we make sure sessionAddress is an `{}`?

File: apps/client/src/pages/QuestionsPage.js
Comment (afjlambert) [inline apps/client/src/pages/QuestionsPage.js]: Can you add a comment?

File: apps/client/src/pages/QuestionsPage.js
Comment (afjlambert) [inline apps/client/src/pages/QuestionsPage.js]: why remove the logic to redirect if the url-slug does not correspond with the question-slug?

File: apps/client/src/pages/QuestionsPage.js
Comment (afjlambert) [inline apps/client/src/pages/QuestionsPage.js]: Can you be a bit more explicit and check answers if the are not `undefined`?

File: apps/client/src/pages/QuestionsPage.js
Comment (afjlambert) [inline apps/client/src/pages/QuestionsPage.js:49]: How can this even work? Value can be `false` right?

File: apps/client/src/pages/QuestionsPage.js
Comment (afjlambert) [inline apps/client/src/pages/QuestionsPage.js:54]: Nice optimization would be to implement the syncronization logic inside the checker itself.
Could be an onSetAnser callback. Or give it a 'store' to put the data in and read from.

File: apps/client/src/pages/QuestionsPage.js
Comment (afjlambert) [inline apps/client/src/pages/QuestionsPage.js]: Checker and checker.stack are guaranteed to be here. Please remove the ?'s.

File: apps/client/src/pages/ResultsPage.js
Comment (afjlambert) [inline apps/client/src/pages/ResultsPage.js]: you can remove the `: questionIndex`.
Doesn't our linter fix this? :-/

File: apps/client/src/sttr_client/models/checker.js
Comment (afjlambert) [inline apps/client/src/sttr_client/models/checker.js]: remove the @returns here please.

File: apps/client/src/sttr_client/models/checker.js
Comment (afjlambert) [inline apps/client/src/sttr_client/models/checker.js]: What?

**Review by afjlambert — CHANGES_REQUESTED** [review]

File: apps/client/src/context.js
Comment (svenjens) [reply apps/client/src/context.js:1]: Let's decide this after the merge with autofill. We might need to refactor some more.

File: apps/client/src/context.js
Comment (svenjens) [inline apps/client/src/context.js:20]: 🙌

File: apps/client/src/hoc/withAddress.js
Comment (svenjens) [inline apps/client/src/hoc/withAddress.js]: Maybe we should check for the topic slug in address here.

File: apps/client/src/hoc/withChecker.js
Comment (svenjens) [reply apps/client/src/hoc/withChecker.js]: We shouldn't need this check right? It's in the withAddress let's check it there.

File: apps/client/src/hoc/withTopic.js
Comment (svenjens) [inline apps/client/src/hoc/withTopic.js:19]: Also, what is resseting.

File: apps/client/src/index.js
Comment (svenjens) [reply apps/client/src/index.js]: it's moved from the bottom to have our import order. 
But i think it should be set on top.

File: apps/client/src/pages/ResultsPage.js
Comment (svenjens) [inline apps/client/src/pages/ResultsPage.js]: let's call this question instead of q

File: apps/client/src/context.js
Comment (robinpiets) [reply apps/client/src/context.js:1]: Good idea, I think it's better to do this after out merge with `autofill`

File: apps/client/src/components/QuestionAnswerTable.js
Comment (robinpiets) [reply apps/client/src/components/QuestionAnswerTable.js]: I don't know why this is. @svenjens why did you add this and can we not remove it? The `sttr-build` is providing us with this information right. For now i've added:
```
} else {
  console.warn(
    `The type "${typeof question.answer}" of "${
      question.answer
    }" is not supported`
  );
}
```

Comment (robinpiets) [reply apps/client/src/components/QuestionAnswerTable.js]: Update: removed this check completely

File: apps/client/src/components/QuestionAnswerTable.js
Comment (robinpiets) [reply apps/client/src/components/QuestionAnswerTable.js]: renamed to `decisivePermits` @afjlambert @svenjens

File: apps/client/src/hoc/withChecker.js
Comment (robinpiets) [reply apps/client/src/hoc/withChecker.js]: @afjlambert `sessionContext.address` is either `null` or an object with data. If `sessionContext.address` exists it means that the address has been set. But as @svenjens  says, it's already been placed in the `withAddress`

File: apps/client/src/index.js
Comment (robinpiets) [reply apps/client/src/index.js]: Indeed, it was at the bottom. I will put it back and will leave it up to the eslint / prettier to deal with import order. For now resolving this

File: apps/client/src/hoc/withAddress.js
Comment (robinpiets) [reply apps/client/src/hoc/withAddress.js]: It's gone now

File: apps/client/src/pages/ConclusionPage.js
Comment (robinpiets) [reply apps/client/src/pages/ConclusionPage.js]: This was actually a bug. It needed `address[slug]`

File: apps/client/src/pages/LocationPage.js
Comment (robinpiets) [reply apps/client/src/pages/LocationPage.js]: This questionmark needs to be there otherwise you get a `TypeError: Cannot read property 'dakkapel-plaatsen' of null` when you open a Question Page in an incognito window. And I don't want to add all slugs to the default. Marked as resolved for now.

File: apps/client/src/pages/LocationPage.js
Comment (robinpiets) [reply apps/client/src/pages/LocationPage.js]: Good point. I was thinking about this as well. But the PR is already going out of control. I will mark this as a separate ticket.

File: apps/client/src/pages/LocationPage.js
Comment (robinpiets) [reply apps/client/src/pages/LocationPage.js]: Solved

File: apps/client/src/pages/QuestionsPage.js
Comment (robinpiets) [reply apps/client/src/pages/QuestionsPage.js]: Code is removed

File: apps/client/src/pages/QuestionsPage.js
Comment (robinpiets) [reply apps/client/src/pages/QuestionsPage.js]: I have no idea, I did not touch this. I will put it back for now. 

@svenjens do you know?

Comment (robinpiets) [reply apps/client/src/pages/QuestionsPage.js]: Fixed: put it back

File: apps/client/src/pages/QuestionsPage.js
Comment (robinpiets) [reply apps/client/src/pages/QuestionsPage.js]: `if (question.options && value !== undefined) {` ?
Or shall we move this to a different ticket?

Comment (robinpiets) [reply apps/client/src/pages/QuestionsPage.js]: Replacing this with value !== undefined is working. Marking as resolved

File: apps/client/src/pages/QuestionsPage.js
Comment (robinpiets) [reply apps/client/src/pages/QuestionsPage.js:49]: On a 'No' answer the value in console.log is `no` and not `false`

File: apps/client/src/pages/QuestionsPage.js
Comment (robinpiets) [reply apps/client/src/pages/QuestionsPage.js:54]: Great!

File: apps/client/src/pages/QuestionsPage.js
Comment (robinpiets) [reply apps/client/src/pages/QuestionsPage.js]: solved

File: apps/client/src/pages/ResultsPage.js
Comment (robinpiets) [reply apps/client/src/pages/ResultsPage.js]: Nice catch

File: apps/client/src/pages/ResultsPage.js
Comment (robinpiets) [reply apps/client/src/pages/ResultsPage.js]: Nice catch, linter did not fix this

File: apps/client/src/sttr_client/models/checker.js
Comment (robinpiets) [reply apps/client/src/sttr_client/models/checker.js]: I had problems with infinite loops. After I added `|| prevId === this._last.id` it seems to have been solved. But I still wanted to keep this as a backup because while (done === false) { is very dangerous.

Remove this fallback?

File: apps/client/src/hoc/withAddress.js
Comment (robinpiets) [reply apps/client/src/hoc/withAddress.js]: Fixed

File: apps/client/src/sttr_client/models/checker.js
Comment (afjlambert) [reply apps/client/src/sttr_client/models/checker.js:111]: Yes, let's remove it.

**Review by afjlambert — APPROVED** [review]

File: apps/client/src/sttr_client/models/checker.js
Comment (robinpiets) [reply apps/client/src/sttr_client/models/checker.js]: Removed (-:

**Pull request merged**
